README
Both sexplib and type-conv have their own build scripts, I would not be removing them, but I would not be using them either.

Run tup upd, then run foobar/foobar and compare the output to this:

$ ./foobar/foobar
foo=((a 1)(b foo))
bar=((a 2)(b bar))


sexplib/Changelog
2011-07-05:  Fixed a parser position bug.  Parser positions passed by the user
             were not updated correctly.

             Internal code beautification.

2011-07-04:  Internal updates to sync with Jane Street.

2011-01-30:  Fixed a code generation bug with toplevel entries.

             Thanks to Yong Lu <lyongu@gmail.com> for the report!

2010-12-27:  Added support for MoreLabels.Hashtbl and improved reporting of
             error locations with preprocessor.

2010-12-26:  Worked around a compiler bug that is expected to be fixed in
             OCaml 3.12.1.  This workaround temporarily removes the interface
             for module Conv (conv.mli), thus exposing the internals.
             This should not cause any problems for end users as long as
             they do not depend on the exported internal representations.
             The interface will become constrained again as soon as the
             fixed compiler is out.

2010-12-22:  Major release.

             Merged with Jane Street version.  This has caused an API-change
             that requires "open Sexplib.Conv" at the top of files that use the
             syntax extension.

             Renamed functions:

               * sexp_of_lazy -> sexp_of_lazy_t
               * lazy_of_sexp -> lazy_t_of_sexp

            Some standard library modules are now re-exported with predefined
            S-expression converters in module Conv.

2010-09-25:  Fixed inferred types of generated functions when dealing with arrow
             types.

2010-08-26:  Fixed a lexer bug when parsing comments.

2010-05-21:  Added support for sexp_bool record field annotations.

2010-05-18:  Improved performance of converting S-expressions to strings.

2010-04-12:  Changed API of Of_sexp_error exception.

2010-04-07:  Added of_(big)string_conv_exn functions.

2010-04-01:  Merged with Jane Street version.

             Major new features (various functions):

               * Type-annotated parsing for better error messages
               * Greatly improved performance of exception converters

2009-12-21:  Improved saving of files.

2009-10-12:  Added sexp_array record field extension.

2009-09-19:  Added missing variant type cases.

             Fixed handling of variance annotations.

2009-09-15:  Internal cleanups.

2009-07-28:  Added better support for conversion of exception types.

2009-06-23:  Fixed build problem.

             Thanks to Sylvain Le Gall <gildor@ocamlcore.org> for
             the patch!

2009-05-08:  Fixed build problems on Windows and OCamlMakefile issues.

             Thanks to Sylvain Le Gall <gildor@ocamlcore.org> for
             the patch!

2009-04-22:  Added macro support for all types of vectors, matrices,
             and for bigstrings.

2009-04-21:  Merged with Jane Street version, no user-relevant changes.

2009-03-09:  Merged with Jane Street version, no user-relevant changes.

2009-03-01:  Fixed build problem on Mac OS X by updating OCamlMakefile.

2009-01-20:  Automatically add S-expression pretty-printers to toplevels.

2008-09-30:  Added a new feature: sexp_opaque.  It prevents the need for /
             use of type converters for a given type in a particular
             type context.

             Removed abstract types to unify them with this new concept.

2008-09-29:  Added a new feature: sexp_list.  This is similar to
             the handling of sexp_option.  By default an empty list is
             assumed for unspecified records using sexp_list as qualifier.
             Such record fields bound to empty lists will also not be
             printed anymore for better readability.

2008-09-23:  Added missing Not_found-exception to standard exception
             converters.

2008-08-20:  Removed dependency on threads.  Fixed build problems.

2008-08-08:  Nifty new feature: exceptions can now be converted to
             S-expressions, too!  The "with sexp" syntax extension can be
             used with exceptions, thus registering a conversion function.
             A global exception conversion function can then be called
             to convert an arbitrary exception into an S-expression,
             which can then be printed out.  This should greatly
             improve readability of uncaught exceptions while making
             life extremely easy for the developer.

             Renamed the ParseError exception to Parse_error to be more
             compliant with Jane Street naming conventions.

2008-07-25:  Added utilities for conversion error handling.  Minor fixes.

2008-04-24:  Made Sexp-interface manifest.

2008-03-20:  Fixed META-file (missing num dependency).

2008-03-17:  Improved META-file.

2008-03-13:  Fully allow function types in converters.  Raise runtime
             exceptions on converting from S-expressions instead when
             function type encountered.

2008-02-11:  Fixed code generation problems with variance annotations
             in signatures, and empty types.

2007-12-17:  Added support for generating signature entries for
             S-expression converters.  Thanks to Till Varoquaux
             <till.varoquaux@gmail.com> for the patch!

2007-11-29:  Added support for converting big_int, nat, num, and ratio.

2007-11-26:  Added support for parsing from bigstrings (char bigarrays).

2007-11-02:  Added syntax support for option types to use the ordinary
             sum type syntax.  This should improve readability.
             The old syntax will be accepted, too, if
             Conv.read_old_option_format is set to true (this is
             currently the default).  The old format will be used for
             writing if Conv.write_old_option_format is true (currently
             the default).  The old syntax is deprecated and will
             probably not be supported by default in the near future.
             Reading new-style option values will always succeed.

2007-09-14:  Fixed bug in S-expression preprocessor concerning record

fieldslib/myocamlbuild.ml
(* OASIS_START *)
(* DO NOT EDIT (digest: e1745ed1bde7b453ebe0f6205cb18764) *)
module OASISGettext = struct
# 21 "/janelibs/ocaml-3.12.0+jane4_64/src/oasis/src/oasis/OASISGettext.ml"
  
  let ns_ str = 
    str
  
  let s_ str = 
    str
  
  let f_ (str : ('a, 'b, 'c, 'd) format4) =
    str
  
  let fn_ fmt1 fmt2 n =
    if n = 1 then
      fmt1^^""
    else
      fmt2^^""
  
  let init = 
    []
  
end

module OASISExpr = struct
# 21 "/janelibs/ocaml-3.12.0+jane4_64/src/oasis/src/oasis/OASISExpr.ml"
  
  
  
  open OASISGettext
  
  type test = string 
  
  type flag = string 
  
  type t =
    | EBool of bool
    | ENot of t
    | EAnd of t * t
    | EOr of t * t
    | EFlag of flag
    | ETest of test * string
    
  
  type 'a choices = (t * 'a) list 
  
  let eval var_get t =
    let rec eval' = 
      function
        | EBool b ->
            b
  
        | ENot e -> 
            not (eval' e)
  
        | EAnd (e1, e2) ->
            (eval' e1) && (eval' e2)
  
        | EOr (e1, e2) -> 
            (eval' e1) || (eval' e2)
  
        | EFlag nm ->
            let v =
              var_get nm
            in
              assert(v = "true" || v = "false");
              (v = "true")
  
        | ETest (nm, vl) ->
            let v =
              var_get nm
            in
              (v = vl)
    in
      eval' t
  
  let choose ?printer ?name var_get lst =
    let rec choose_aux = 
      function
        | (cond, vl) :: tl ->
            if eval var_get cond then 
              vl 
            else
              choose_aux tl
        | [] ->
            let str_lst = 
              if lst = [] then
                s_ "<empty>"
              else
                String.concat 
                  (s_ ", ")
                  (List.map
                     (fun (cond, vl) ->
                        match printer with
                          | Some p -> p vl
                          | None -> s_ "<no printer>")
                     lst)
            in
              match name with 
                | Some nm ->
                    failwith
                      (Printf.sprintf 
                         (f_ "No result for the choice list '%s': %s")
                         nm str_lst)
                | None ->
                    failwith
                      (Printf.sprintf
                         (f_ "No result for a choice list: %s")
                         str_lst)
    in
      choose_aux (List.rev lst)
  
end


module BaseEnvLight = struct
# 21 "/janelibs/ocaml-3.12.0+jane4_64/src/oasis/src/base/BaseEnvLight.ml"
  
  module MapString = Map.Make(String)
  
  type t = string MapString.t
  
  let default_filename =
    Filename.concat 
      (Sys.getcwd ())
      "setup.data"
  
  let load ?(allow_empty=false) ?(filename=default_filename) () =
    if Sys.file_exists filename then
      begin
        let chn =
          open_in_bin filename
        in
        let st =
          Stream.of_channel chn
        in
        let line =
          ref 1
        in
        let st_line = 
          Stream.from
            (fun _ ->
               try
                 match Stream.next st with 
                   | '\n' -> incr line; Some '\n'
                   | c -> Some c
               with Stream.Failure -> None)
        in
        let lexer = 
          Genlex.make_lexer ["="] st_line
        in
        let rec read_file mp =
          match Stream.npeek 3 lexer with 
            | [Genlex.Ident nm; Genlex.Kwd "="; Genlex.String value] ->
                Stream.junk lexer; 
                Stream.junk lexer; 
                Stream.junk lexer;
                read_file (MapString.add nm value mp)
            | [] ->
                mp
            | _ ->
                failwith
                  (Printf.sprintf
                     "Malformed data file '%s' line %d"
                     filename !line)
        in
        let mp =
          read_file MapString.empty
        in
          close_in chn;
          mp
      end
    else if allow_empty then
      begin
        MapString.empty
      end
    else
      begin
        failwith 
          (Printf.sprintf 
             "Unable to load environment, the file '%s' doesn't exist."
             filename)
      end
  
  let var_get name env =
    let rec var_expand str =
      let buff =
        Buffer.create ((String.length str) * 2)
      in
        Buffer.add_substitute 
          buff
          (fun var -> 
             try 
               var_expand (MapString.find var env)
             with Not_found ->
               failwith 
                 (Printf.sprintf 
                    "No variable %s defined when trying to expand %S."
                    var 
                    str))
          str;
        Buffer.contents buff
    in
      var_expand (MapString.find name env)
  
  let var_choose lst env = 
    OASISExpr.choose
      (fun nm -> var_get nm env)
      lst
end


module MyOCamlbuildFindlib = struct
# 21 "/janelibs/ocaml-3.12.0+jane4_64/src/oasis/src/plugins/ocamlbuild/MyOCamlbuildFindlib.ml"
  
  (** OCamlbuild extension, copied from 
    * http://brion.inria.fr/gallium/index.php/Using_ocamlfind_with_ocamlbuild
    * by N. Pouillard and others
    *
    * Updated on 2009/02/28
    *
    * Modified by Sylvain Le Gall 
    *)
  open Ocamlbuild_plugin
  
  (* these functions are not really officially exported *)
  let run_and_read = 
    Ocamlbuild_pack.My_unix.run_and_read
  
  let blank_sep_strings = 
    Ocamlbuild_pack.Lexers.blank_sep_strings
  
  let split s ch =
    let x = 
      ref [] 
    in
    let rec go s =
      let pos = 
        String.index s ch 
      in
        x := (String.before s pos)::!x;
        go (String.after s (pos + 1))
    in
      try
        go s
      with Not_found -> !x
  
  let split_nl s = split s '\n'
  
  let before_space s =
    try
      String.before s (String.index s ' ')
    with Not_found -> s
  
  (* this lists all supported packages *)
  let find_packages () =
    List.map before_space (split_nl & run_and_read "ocamlfind list")
  
  (* this is supposed to list available syntaxes, but I don't know how to do it. *)
  let find_syntaxes () = ["camlp4o"; "camlp4r"]
  
  (* ocamlfind command *)
  let ocamlfind x = S[A"ocamlfind"; x]
  
  let dispatch =
    function
      | Before_options ->
          (* by using Before_options one let command line options have an higher priority *)
          (* on the contrary using After_options will guarantee to have the higher priority *)
          (* override default commands by ocamlfind ones *)
          Options.ocamlc     := ocamlfind & A"ocamlc";
          Options.ocamlopt   := ocamlfind & A"ocamlopt";
          Options.ocamldep   := ocamlfind & A"ocamldep";
          Options.ocamldoc   := ocamlfind & A"ocamldoc";
          Options.ocamlmktop := ocamlfind & A"ocamlmktop"
                                  
      | After_rules ->
          
          (* When one link an OCaml library/binary/package, one should use -linkpkg *)
          flag ["ocaml"; "link"; "program"] & A"-linkpkg";
          
          (* For each ocamlfind package one inject the -package option when
           * compiling, computing dependencies, generating documentation and
           * linking. *)
          List.iter 
            begin fun pkg ->
              flag ["ocaml"; "compile";  "pkg_"^pkg] & S[A"-package"; A pkg];
              flag ["ocaml"; "ocamldep"; "pkg_"^pkg] & S[A"-package"; A pkg];
              flag ["ocaml"; "doc";      "pkg_"^pkg] & S[A"-package"; A pkg];
              flag ["ocaml"; "link";     "pkg_"^pkg] & S[A"-package"; A pkg];
              flag ["ocaml"; "infer_interface"; "pkg_"^pkg] & S[A"-package"; A pkg];
            end 
            (find_packages ());
  
          (* Like -package but for extensions syntax. Morover -syntax is useless
           * when linking. *)
          List.iter begin fun syntax ->
          flag ["ocaml"; "compile";  "syntax_"^syntax] & S[A"-syntax"; A syntax];
          flag ["ocaml"; "ocamldep"; "syntax_"^syntax] & S[A"-syntax"; A syntax];
          flag ["ocaml"; "doc";      "syntax_"^syntax] & S[A"-syntax"; A syntax];
          flag ["ocaml"; "infer_interface"; "syntax_"^syntax] & S[A"-syntax"; A syntax];
          end (find_syntaxes ());
  
          (* The default "thread" tag is not compatible with ocamlfind.
           * Indeed, the default rules add the "threads.cma" or "threads.cmxa"
           * options when using this tag. When using the "-linkpkg" option with
           * ocamlfind, this module will then be added twice on the command line.
           *                        
           * To solve this, one approach is to add the "-thread" option when using
           * the "threads" package using the previous plugin.
           *)
          flag ["ocaml"; "pkg_threads"; "doc"] (S[A "-thread"]);
          flag ["ocaml"; "pkg_threads"; "compile"] (S[A "-thread"]);
          flag ["ocaml"; "pkg_threads"; "link"] (S[A "-thread"]);
          flag ["ocaml"; "pkg_threads"; "infer_interface"] (S[A "-thread"])
  
      | _ -> 
          ()
  
end

module MyOCamlbuildBase = struct
# 21 "/janelibs/ocaml-3.12.0+jane4_64/src/oasis/src/plugins/ocamlbuild/MyOCamlbuildBase.ml"
  
  (** Base functions for writing myocamlbuild.ml
      @author Sylvain Le Gall
    *)
  
  
  
  open Ocamlbuild_plugin
  
  type dir = string 
  type file = string 
  type name = string 
  type tag = string 
  
# 55 "/janelibs/ocaml-3.12.0+jane4_64/src/oasis/src/plugins/ocamlbuild/MyOCamlbuildBase.ml"
  
  type t =
      {
        lib_ocaml: (name * dir list) list;
        lib_c:     (name * dir * file list) list; 
        flags:     (tag list * (spec OASISExpr.choices)) list;
      } 
  
  let env_filename =
    Pathname.basename 
      BaseEnvLight.default_filename
  
  let dispatch_combine lst =
    fun e ->
      List.iter 
        (fun dispatch -> dispatch e)
        lst 
  
  let dispatch t e = 
    let env = 
      BaseEnvLight.load 
        ~filename:env_filename 
        ~allow_empty:true
        ()
    in
      match e with 
        | Before_options ->
            let no_trailing_dot s =
              if String.length s >= 1 && s.[0] = '.' then
                String.sub s 1 ((String.length s) - 1)
              else
                s
            in
              List.iter
                (fun (opt, var) ->
                   try 
                     opt := no_trailing_dot (BaseEnvLight.var_get var env)
                   with Not_found ->
                     Printf.eprintf "W: Cannot get variable %s" var)
                [
                  Options.ext_obj, "ext_obj";
                  Options.ext_lib, "ext_lib";
                  Options.ext_dll, "ext_dll";
                ]
  
        | After_rules -> 
            (* Declare OCaml libraries *)
            List.iter 
              (function
                 | lib, [] ->
                     ocaml_lib lib;
                 | lib, dir :: tl ->
                     ocaml_lib ~dir:dir lib;
                     List.iter 
                       (fun dir -> 
                          flag 
                            ["ocaml"; "use_"^lib; "compile"] 
                            (S[A"-I"; P dir]))
                       tl)
              t.lib_ocaml;
  
            (* Declare C libraries *)
            List.iter
              (fun (lib, dir, headers) ->
                   (* Handle C part of library *)
                   flag ["link"; "library"; "ocaml"; "byte"; "use_lib"^lib]
                     (S[A"-dllib"; A("-l"^lib); A"-cclib"; A("-l"^lib)]);
  
                   flag ["link"; "library"; "ocaml"; "native"; "use_lib"^lib]
                     (S[A"-cclib"; A("-l"^lib)]);
                        
                   flag ["link"; "program"; "ocaml"; "byte"; "use_lib"^lib]
                     (S[A"-dllib"; A("dll"^lib)]);
  
                   (* When ocaml link something that use the C library, then one
                      need that file to be up to date.
                    *)
                   dep  ["compile"; "ocaml"; "use_lib"^lib] 
                     [dir/"lib"^lib^"."^(!Options.ext_lib)];
  
                   (* TODO: be more specific about what depends on headers *)
                   (* Depends on .h files *)
                   dep ["compile"; "c"] 
                     headers;
  
                   (* Setup search path for lib *)
                   flag ["link"; "ocaml"; "use_"^lib] 
                     (S[A"-I"; P(dir)]);
              )
              t.lib_c;
  
              (* Add flags *)
              List.iter
              (fun (tags, cond_specs) ->
                 let spec = 
                   BaseEnvLight.var_choose cond_specs env
                 in
                   flag tags & spec)
              t.flags
        | _ -> 
            ()
  
  let dispatch_default t =
    dispatch_combine 
      [
        dispatch t;
        MyOCamlbuildFindlib.dispatch;
      ]
  
end


open Ocamlbuild_plugin;;
let package_default =
  {
     MyOCamlbuildBase.lib_ocaml =
       [("lib/fieldslib", ["lib"]); ("syntax/pa_fields_conv", ["syntax"])];
     lib_c = [];
     flags = [];
     }
  ;;

let dispatch_default = MyOCamlbuildBase.dispatch_default package_default;;

(* OASIS_STOP *)
Ocamlbuild_plugin.dispatch dispatch_default;;
